### ISSUE TITLE
[BUG] TLC operator '@@' for function merge does not work as expected.

### ISSUE THREAD

**Issue by danwt** [informalsystems/apalache#951]
## Description

TLC operator '@@' for function merge does not work as expected. 

The TLC version is as follows

```
f @@ g == [x \in (DOMAIN f) \cup (DOMAIN g) |-> IF x \in DOMAIN f THEN f[x] ELSE g[x]]
```

The behavior of preferring `f` in case of domain intersect is not respected by Apalache.

## Input specification

```
------------------------------------ MODULE main -----------------------------
EXTENDS Apalache, Integers, FiniteSets, Sequences, TLC
VARIABLES 
    \* @type: Int -> Int;
    myfun

Init ==
    myfun = [e \in 1..3 |-> e]

Next ==
    myfun' = myfun @@ [e \in 2..4 |-> 42]

MyInv == 
    /\ myfun[2] = 2
    /\ myfun[3] = 3
===============================================================================
```

## The command line parameters used to run the tool

`apalache-mc check --inv=MyInv main.tla`

## Expected behavior

No error

## Log files

```
State 1: state invariant 0 violated. Check the counterexample in: counterexample1.tla, MC1.out, counterexample1.json E@09:26:02.564
Found 1 error(s)                                                  I@09:26:02.566
The outcome is: Error                                             I@09:26:02.567
Checker has found an error
```

## System information

- Apalache version [`apalache-mc version`]: APALACHE version 0.15.12 build 05d50b2
- OS [e.g. Ubuntu Linux or Mac OS, and the current version]: macOS 11.5, build 20G71
- JDK version [e.g. OpenJDK 0.8.0]: OpenJDK Runtime Environment Zulu16.32+15-CA (build 16.0.2+7)

## Additional context

Suspected relevant Apalache src location:

`apalache/tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/TlcRule.scala` method `extendFun`

```
  private def extendFun(state: SymbState, leftFunEx: TlaEx, rightFunEx: TlaEx): SymbState = {
    def solverAssert = rewriter.solverContext.assertGroundExpr _

    val funT1 = TlaType1.fromTypeTag(leftFunEx.typeTag).asInstanceOf[FunT1]
    var nextState = rewriter.rewriteUntilDone(state.setRex(leftFunEx))
    val leftFunCell = nextState.asCell
    nextState = rewriter.rewriteUntilDone(nextState.setRex(rightFunEx))
    val rightFunCell = nextState.asCell
    **// Blindly concatenate both relations. If the domains of the functions intersect, we may produce an unsound encoding.
    // As TLC!@@ is used only to decode counterexamples, it should be ok. Otherwise, we would produce a lot of** constraints.
    val leftRelation = nextState.arena.getCdm(leftFunCell)
    val leftPairs = nextState.arena.getHas(leftRelation)
    val rightRelation = nextState.arena.getCdm(rightFunCell)
    val rightPairs = nextState.arena.getHas(rightRelation)
    val jointPairs = leftPairs ++ rightPairs
    nextState = nextState.updateArena(_.appendCell(leftFunCell.cellType))
    val newFunCell = nextState.arena.topCell
    nextState = nextState.updateArena(_.appendCell(leftRelation.cellType))
    val newRelation = nextState.arena.topCell
    nextState = nextState.updateArena(_.setCdm(newFunCell, newRelation)
          .appendHas(newRelation, jointPairs: _*))

    // As we pass the expressions to SMT, we could use untyped expressions.
    // We don't do it, in order to avoid mixing untyped and typed expressions in the same class.
    val pairT = TupT1(funT1.arg, funT1.res)
    val types =
      Map("b" -> BoolT1(), "p" -> pairT, "r" -> SetT1(pairT))
    // the new pair unconditionally belongs to the new cell
    for (pair <- leftPairs) {
      val inOld = tla.in(pair.toNameEx ? "p", leftRelation.toNameEx ? "r") ? "b"
      val inNew = tla.in(pair.toNameEx ? "p", newRelation.toNameEx ? "r") ? "b"
      solverAssert(tla.equiv(inNew, inOld).typed(types, "b"))
    }
    for (pair <- rightPairs) {
      val inOld = tla.in(pair.toNameEx ? "p", rightRelation.toNameEx ? "r") ? "b"
      val inNew = tla.in(pair.toNameEx ? "p", newRelation.toNameEx ? "r") ? "b"
      solverAssert(tla.equiv(inNew, inOld).typed(types, "b"))
    }

    nextState.setRex(newFunCell.toNameEx)
  }
```

This operator can be useful for avoiding writing EXCEPT everywhere.

**Comment by danwt** [comment]
I can try to fix this in a few months - it would likely be a good starting point to get into the guts.

**Comment by konnov** [comment]
Nice catch. I did not pay attention to the exact definition by TLC. I would rather expect it to use `CHOOSE` in the case when an element belongs to the intersection, which would make the operator symmetric. I agree, it would be nice to fix it when you have time.

**Comment by konnov** [comment]
Actually, I realized that this definition of `@@` does not agree with the common definition of `Map.add` or `Map.put` in the programming languages that I know. Not that I am surprised. But if you want to maintain the "standard" PL approach with `@@`, you would have to put a new pair on the left.

**Comment by danwt** [comment]
Exactly, it's the opposite of what you would expect.

**Comment by konnov** [comment]
Just for the note, the test in this issue is failing the typechecker in #1253. To be fixed when #1253 is merged.

**Comment by konnov** [comment]
This one is working now
